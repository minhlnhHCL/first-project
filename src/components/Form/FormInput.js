import React from 'react'

const FormInput = ({
    inputName,
    inputLabel,
    inputValue,
    inputType = 'text',
    inputError,
    handleChange
}) => {
    return (
        <div className='d-flex-ac mt-32'>
            <label htmlFor={inputName} className="custom-label">
                {inputLabel}
            </label>
            <inpu
                type={inputType}
                className="custom-input"
                id={inputName}
                name={inputName}
                value={inputValue}
                onChange={handleChange}
            />
        </div>
    )
}

export default FormInput