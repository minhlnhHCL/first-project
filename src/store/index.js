import { configureStore } from '@reduxjs/toolkit'
import { combineReducers } from 'redux'
import users from './userSlice'

const reducer = combineReducers({
    users
})
const setupStore = preloadedState => {
    return configureStore({
        reducer,
        preloadedState
    })
}
export default setupStore;