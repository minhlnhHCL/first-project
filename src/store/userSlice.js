import { createSlice } from '@reduxjs/toolkit'
export const GET_USER = 'GET_USER'
export const ADD_USER = 'ADD_USER'
export const EDIT_USER = 'EDIT_USER'
export const DELETE_USER = 'ADELETEUSER'
export const CURRENT_USER = 'CURRENT_USER'

export const initialState = {
    users: [],
    currentUser: {
        id: 0,
        firstName: '',
        lastName: '',
        image: '',
        age: 0,
        company: {
            title: ''
        }
    }
}
//UserSlice

export const userSLice = createSlice({
    name: 'users',
    initialState,
    reducers: {
        getUsers: (state, action) => {
            if (state.users.length == 1) {
                const newUserArr = [...state.users, ...action.payload]
                state.users = newUserArr
                return state
            }
            state.users = action.payload
        },
        addUser: (state, action) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            console.log('index', index)

            if (index == -1) {
                console.log('vo day')
                state.users = [action.payload, ...state.users]
                return state
            }
            return state
        },
        setCurrentUser: (state, action) => {
            state.currentUser = action.payload
        },
        editUser: (state, action) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            const newUserArr = [...state.users]
            newUserArr[index] = action.payload
            state.users = newUserArr
        },
        deleteUser: (state, action) => {
            const clonedArr = [...state.users]
            const newUserArr = clonedArr.filter(item => item.id !== action.payload.id)
            state.users = newUserArr
        }
    }
})

export const { getUsers, editUser, addUser, deleteUser, setCurrentUser } = userSLice.actions
export default userSLice.reducer