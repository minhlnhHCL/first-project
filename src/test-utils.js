import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { BrowserRouter, Routes } from "react-router-dom"
import { SpinnerContext, SpinnerProvider } from "./context/SpinnerContext"
import setupStore from "./store"

// export const renderWithRedux = (children) => {
//     const renderResult = render(<Provider store={store}>
//         <SpinnerProvider value={SpinnerContext}>
//             {children}
//         </SpinnerProvider>
//     </Provider>)
//     return {
//         ...renderResult,
//         store,
//     };
// }
// re-export everything
export * from '@testing-library/react'

export const renderWithRedux = (
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    } = {}
) => {
    const Wrapper = ({ children }) => {
        return (
            <Provider store={store}>
                <SpinnerProvider value={SpinnerContext}>
                    {children}</SpinnerProvider>
            </Provider>
        )
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}