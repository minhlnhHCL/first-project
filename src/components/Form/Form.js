
import axios from 'axios'
import React, { useContext, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { SpinnerContext } from '../../context/SpinnerContext'
import { addUser, editUser, setCurrentUser } from '../../store/userSlice'
import { useDispatch, useSelector } from 'react-redux'
import './Form.scss'
const Form = () => {
  const { users, currentUser } = useSelector(state => state.users)
  const dispatch = useDispatch()
  const { onLoading, offLoading } = useContext(SpinnerContext)
  const [firstName, setFirstName] = useState(currentUser.firstName)
  const [lastName, setLastName] = useState(currentUser.lastName)
  const [companyTitle, setCompanyTitle] = useState(currentUser.company.title)
  const navigate = useNavigate()

  const onSubmit = async (event) => {
    event.preventDefault()
    onLoading()
    try {
      const body = { ...currentUser }
      body.firstName = firstName
      body.lastName = lastName
      body.company = { ...body.company, title: companyTitle }
      if (currentUser.id) {
        const req = await axios.put(`https://dummyjson.com/users/${currentUser.id}`, body)
        if (req.status == 200) {
          dispatch(editUser(body))
          offLoading()
          return navigate('/')

        }
      }
      const newId = users.length == 0 ? 31 : users[users.length == 31 ? 0 : users.length - 1].id + 1
      body.id = newId
      const req = await axios.post('https://dummyjson.com/users/add', body)
      if (req.status == 200) {
        dispatch(addUser(body))
        offLoading()
        return navigate('/')
      }
    } catch (error) {
      offLoading()
      alert(error.message);
    }
  }

  return (
    <form onSubmit={onSubmit}>
      <div className='d-flex-ac'>
        <label className='custom-label'>First Name: </label>
        <input className='custom-input ' placeholder='First Name' name='firstName' value={firstName} onChange={(e) => setFirstName(e.target.value)} />
      </div>
      <div className='d-flex-ac mt-32'>
        <label className='custom-label'>Last Name: </label>
        <input className='custom-input' placeholder='Last Name' name='lastName' value={lastName} onChange={(e) => setLastName(e.target.value)} />
      </div>
      <div className='d-flex-ac mt-32'>
        <label className='custom-label'>Last Name: </label>
        <input className='custom-input' placeholder='Company Title' name='companyTitle' value={companyTitle} onChange={(e) => setCompanyTitle(e.target.value)} />
      </div>

      <div className='form-actions'>
        <div className="action-container">
          <button className='btn cancel' type='button' onClick={() => {
            dispatch(setCurrentUser({ firstName: '', lastName: '', age: 0, id: 0, company: { title: '' } }))
          }}>
            Cancel
          </button>
          <button aria-label='submit' name='btn-submit' className='btn submit'disabled={!(firstName && lastName && companyTitle)} type='submit'>Submit</button>
        </div>
      </div>
    </form >
  )
}

export default Form