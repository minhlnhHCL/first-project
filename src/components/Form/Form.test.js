import { useDispatch, useSelector } from "react-redux";
import store from "../../store";
import axios from "axios";
import Form from "./Form";
import {
  cleanup,
  render,
  waitForElementToBeRemoved,
  screen,
  waitFor,
} from "@testing-library/react";
import { useNavigate } from "react-router-dom";
import { renderWithRedux } from "../../test-utils";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.mock("axios");

describe("User Form", () => {
  afterAll(cleanup);
  beforeEach(() => {
    useDispatch.mockReturnValue(() => { });
    useSelector.mockReturnValue({
      currentUser: {
        firstName: '',
        lastName: '',
        image: '',
        age: 0,
        company: {
          title: ''
        }
      }
    });
  });

  it("should display empty form for new user", () => {
    const { container } = renderWithRedux(<Form />, {
      preloadedState: {
        currentUser: {
          firstName: '',
          lastName: '',
          image: '',
          age: 0,
          company: {
            title: ''
          }
        }
      }
    });
    expect(container).toMatchSnapshot();
  });
});
