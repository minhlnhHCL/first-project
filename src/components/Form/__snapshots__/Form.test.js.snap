// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Form should display empty form for new user 1`] = `
<div>
  <form>
    <div
      class="d-flex-ac"
    >
      <label
        class="custom-label"
      >
        First Name: 
      </label>
      <input
        class="custom-input "
        name="firstName"
        placeholder="First Name"
        value=""
      />
    </div>
    <div
      class="d-flex-ac mt-32"
    >
      <label
        class="custom-label"
      >
        Last Name: 
      </label>
      <input
        class="custom-input"
        name="lastName"
        placeholder="Last Name"
        value=""
      />
    </div>
    <div
      class="d-flex-ac mt-32"
    >
      <label
        class="custom-label"
      >
        Last Name: 
      </label>
      <input
        class="custom-input"
        name="companyTitle"
        placeholder="Company Title"
        value=""
      />
    </div>
    <div
      class="form-actions"
    >
      <div
        class="action-container"
      >
        <button
          class="btn cancel"
          type="button"
        >
          Cancel
        </button>
        <button
          aria-label="submit"
          class="btn submit"
          disabled=""
          name="btn-submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
